core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -174,6 +177,24 @@ private[spark] class CoarseGrainedExecutorBackend(
 
 private[spark] object CoarseGrainedExecutorBackend extends Logging {
 
+  private def addMesosDelegationTokens(driverConf: SparkConf) {
+    val value = driverConf.get("spark.mesos.kerberos.userCredentials")
+    val tokens = DatatypeConverter.parseBase64Binary(value)
+
+    logDebug(s"Found delegation tokens of ${tokens.length} bytes.")
+
+    // Use tokens for HDFS login.
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(driverConf)
+    hadoopConf.set("hadoop.security.authentication", "Token")
+    UserGroupInformation.setConfiguration(hadoopConf)
+
+    // Decode tokens and add them to the current user's credentials.
+    val creds = UserGroupInformation.getCurrentUser.getCredentials
+    val tokensBuf = new java.io.ByteArrayInputStream(tokens)
+    creds.readTokenStorageStream(new java.io.DataInputStream(tokensBuf))
+    UserGroupInformation.getCurrentUser.addCredentials(creds)
+  }
```

Comment:
To elaborate; one potential use of it is to do token acquisition and token distribution without needing to provide principal/keytab to spark application (other than keeping track of last credential update to account for AM failures).
This is WIP though, but if mesos approach is different from yarn - that would be a great way to iterate on the latter aspect (token distribution); and ensure it is extensible enough for future requirements/implementations.